server/src/main/java/org/elasticsearch/action/support/TransportActions.java

```diff
@@ -35,4 +47,48 @@ public static boolean isReadOverrideException(Exception e) {
         return isShardNotAvailableException(e) == false;
     }
 
+    /**
+     * Broadcasts a given action request to all unpromotable assigned shards of a provided {@link ShardId}.
+     *
+     * @param shardId                        the routing table of this given shard will be used to get all unpromotable assigned shards
+     * @param request                        the request to broadcast
+     * @param action                         the action to broadcast
+     * @param parentTask                     the task that will be parent of the children broadcast requests
+     * @param clusterService                 the cluster service
+     * @param transportService               the transport service
+     * @param responseHandlerExecutor        the executor to be used in the {@link ActionListenerResponseHandler}
+     * @param listener                       the listener to notify when all children requests are done
+     */
+    public static void broadcastToUnpromotableShards(
+        ShardId shardId,
+        ActionRequest request,
+        String action,
+        Task parentTask,
+        ClusterService clusterService,
+        TransportService transportService,
+        String responseHandlerExecutor,
+        ActionListener<Void> listener
+    ) {
```

Comment:
I am new to this area, but I wonder if this should be a static utility?
It seems like there is a pattern of making abstract actions like `TransportBroadcastAction`.
May be @tlrx or @DaveCTurner could advice on this.

Reply:
Written like this, I would agree. However, I think the caller should be able to pass in an `IndexShardRoutingTable` rather than having to read the cluster state from the `ClusterService`, because the state in the `ClusterService` may differ from the one that the caller expects. With that change, it's a bit harder to express this as another kind of abstract action - the `IndexShardRoutingTable` would have to be carried in the request itself, which must be an `ActionRequest`, and hence `Writeable`. Maybe that's ok? It still seems a bit weird cos we never actually want to send such a request over the wire.

Reply:
Also do we want to test this logic or it is okay that it is implicitly covered by other tests?

Reply:
> Also do we want to test this logic or it is okay that it is implicitly covered by other tests?

For the moment, this is tested in org.elasticsearch.cluster.routing.ShardRoutingRoleIT#testRefreshOfUnpromotableShards indirectly. That is the only IT file so far in ES that tests unpromotable shards. I will see if/how I can make a generic test for the new utility function there.

Reply:
> we cannot make an abstract action as Refresh cannot inherit two classes.

That's ok tho, the unpromotable bit would be a separate action.

> But I believe I would still need the ClusterService to get the cluster state to convert shards' node IDs to DiscoveryNodes, right?

Ah ok you'd need to accept the `DiscoveryNodes` too.

> Maybe that's ok?

I've thought about this for a bit and decided that yes this is ok and indeed I prefer it 😄

Reply:
> I've thought about this for a bit and decided that yes this is ok and indeed I prefer it

Trying to connect the dots, you mean we can make a new abstract action that underneaths broadcasts it to unpromotable shards, and Refresh does not inherit it, but rather executes it? I guess that will work and is in line with @idegtiarenko 's comment. Let me see if/how I can make this.

BTW, we'd still need the hook of this PR for the custom logic so that Refresh actually shoots the new abstract action.

Reply:
> Refresh does not inherit it, but rather executes it

Yep.

> we'd still need the hook of this PR for the custom logic

Yep.

> I will see if/how I can make a generic test for the new utility function there.

I would prefer to have unit tests covering this logic. It is a little awkward to set everything up for such a test, but it's well worth it. See e.g. `TransportNodesActionTests` or `TransportBroadcastByNodeActionTests` for some test suites of similar things.